test/functional/cursor.test.js

```diff
@@ -4372,6 +4372,72 @@ describe('Cursor', function() {
     }
   );
 
+  describe('Cursor forEach Error propagation', function() {
+    let configuration;
+    let client;
+    let cursor;
+    let collection;
+
+    beforeEach(function(done) {
+      configuration = this.configuration;
+      client = configuration.newClient({ w: 1 }, { maxPoolSize: 1 });
+      client
+        .connect()
+        .then(() => {
+          collection = client.db(configuration.db).collection('cursor_session_tests2');
+          done();
+        })
+        .catch(error => {
+          done(error);
+        });
+    });
+
+    afterEach(function(done) {
+      if (cursor) {
+        cursor
+          .close()
+          .then(() => client.close())
+          .then(() => done());
+      } else {
+        client.close().then(() => done());
+      }
+    });
+
+    // NODE-2035
+    it('should propagate error when exceptions are thrown from an awaited forEach call', function(done) {
+      const docs = [{ unique_key_2035: 1 }, { unique_key_2035: 2 }, { unique_key_2035: 3 }];
+      collection.insertMany(docs).then(
+        () => {
+          cursor = collection.find({
+            unique_key_2035: {
+              $exists: true
+            }
+          });
+          cursor
+            .forEach(() => {
+              throw new Error('FAILURE IN FOREACH CALL');
+            })
+            .then(
+              () => {
+                done(new Error('Error in forEach call not caught'));
+              },
+              err => {
+                try {
+                  expect(err.message).to.deep.equal('FAILURE IN FOREACH CALL');
+                  done();
+                } catch (error) {
+                  done(error);
+                }
+              }
+            );
+        },
+        error => {
```

Comment:
Instead of having the error here, I recommend moving it down to a .catch block; that would ensure that if the .forEach throws synchronously in L4417, that error will also be caught and passed to the done callback; if you keep the error handler as you do now on the top level then, it will only handle the error from insertMany(), and there would be nothing to handle sync errors from the success handler (because the L4422 error handler is only for explicit rejections on the forEach, not for any uncaught sync errors that could arise in the execution of the .forEach handler itself). Does that make sense?

Reply:
Just tried out adding a throw in the forEach implementation itself, so that makes perfect sense. Thanks so much for pointing that out. Wouldn't have seen it otherwise. 👍🏿